File: nodejs/service.js
import { persistentToFile, readFromFile } from './store.js'

async function handleGetTodoList(status) {
    const todoList = await readFromFile()
    if (['active', 'completed'].includes(status)) {
        return todoList.filter(it => it.status === status)
    }
    return todoList
}

async function handleAddNewTodo(content) {
    if (!content) {
        throw new Error('Todo content must be not empty')
    }
    const todoList = await readFromFile()
    const newId = todoList.length ? Math.max(...todoList.map(it => it.id)) + 1 : 1
    const newTodo = {
        id: newId,
        content,
        status: 'active',
    }
    todoList.push(newTodo)
    await persistentToFile(todoList)
    return newTodo
}

async function handleUpdateTodo(todo) {
    if (!todo.id) {
        throw new Error('Todo ID is missing')
    }
    if (!['active', 'completed'].includes(todo.status)) {
        throw new Error('Illegal todo status')
    }
    if (!todo.content) {
        throw new Error('Todo content must be not empty')
    }
    const todoList = await readFromFile()
    const newTodoList = todoList.map(it => it.id === todo.id ? todo : it)
    await persistentToFile(newTodoList)
}

async function handleBulkDeleteTodo(idList) {
    if (!idList.length) {
        return
    }
    const todoList = await readFromFile()
    const newTodoList = todoList.filter(it => !idList.includes(it.id))
    await persistentToFile(newTodoList)
}

function parseJsonBody(req) {
    return new Promise(((resolve, reject) => {
        let data = '';
        req.on('data', chunk => {
            data += chunk;
        })
        req.on('end', () => {
            resolve(JSON.parse(data))
        })
        req.on('error', (err) => {
            reject(err)
        })
    }))
}

export default async function service(req, res) {
    try {
        const url = new URL(req.url, 'http://service/')
        const route = `${req.method} ${url.pathname}`
        switch (route) {
            case 'GET /':
                res.end('Todo MVC Service')
                break
            case 'GET /todo/list':
                const status = url.searchParams.get('status') ?? ''
                const todoList = await handleGetTodoList(status)
                res.setHeader('Content-Type', 'application/json')
                res.end(JSON.stringify(todoList))
                break
            case 'POST /todo':
                const { content } = await parseJsonBody(req) ?? {}
                const newTodo = await handleAddNewTodo(content)
                res.setHeader('Content-Type', 'application/json')
                res.end(JSON.stringify(newTodo))
                break
            case 'PUT /todo':
                const todo = await parseJsonBody(req)
                await handleUpdateTodo(todo)
                res.end()
                break
            case 'DELETE /todo':
                const idList = url.searchParams.getAll('id')
                    .map(id => Number(id))
                    .filter(id => !!id)
                await handleBulkDeleteTodo(idList)
                res.end()
                break
            default:
                res.statusCode = 404
                res.end('404 Not Found')
        }
    } catch (err) {
        res.statusCode = 400
        res.end('400 ' + err.message ?? 'Bad Request')
    }
}

File: ruby/service.rb
require 'webrick'
require 'json'
require_relative 'store'

class TodoListHandler < WEBrick::HTTPServlet::AbstractServlet

    def do_GET request, response
        status = request.query['status']

        response.status = 200
        response['Content-Type'] = 'application/json'
        response.body = self.get_todo_list status
    end

    def get_todo_list status
        todo_list = read_from_file
        if status === 'active' || status === 'completed'
            todo_list = todo_list.select do |it| it['status'] === status end
        end
        return JSON.generate todo_list
    end
end

class TodoHandler < WEBrick::HTTPServlet::AbstractServlet

    def do_POST request, response
        payload = JSON.parse(request.body)
        new_todo = self.add_new_todo payload['content']

        response.status = 200
        response['Content-Type'] = 'text/plain'
        response.body = JSON.generate new_todo
    end

    def do_PUT request, response
        payload = JSON.parse(request.body)
        self.update_todo payload

        response.status = 200
        response['Content-Type'] = 'text/plain'
        response.body = ''
    end

    def do_DELETE request, response
        id_list = self.parse_id_List(request.query_string)
        self.delete_todo id_list

        response.status = 200
        response['Content-Type'] = 'text/plain'
        response.body = ''
    end

    def add_new_todo content
        todo_list = read_from_file
        new_id = 1
        if(!todo_list.empty?)
           id_list = todo_list.map do |it| it['id'] end
           new_id = id_list.max + 1
        end
        new_todo = { :id => new_id, :content => content, :status => 'active' }
        todo_list << new_todo
        persistent_to_file todo_list
        return new_todo
    end

    def update_todo todo
        todo_list = read_from_file
        new_todo_list = todo_list.map do |it|
            if(it['id'] === todo['id'])
                { :id => todo['id'], :content => todo['content'], :status => todo['status']}
            else
                it
            end
        end
        persistent_to_file new_todo_list
    end

    def delete_todo id_list
        todo_list = read_from_file
        new_todo_list = todo_list.select do |it| !id_list.include? it['id'] end
        persistent_to_file new_todo_list
    end

    def parse_id_List query_string
        if query_string === ''
            return []
        end
        return (query_string.split '&').map do |it| (it.split '=')[1].to_i end
    end

end

def create_server(port)
    server = WEBrick::HTTPServer.new(:Port => port)

    server.mount_proc '/' do |req, res|
        res.body = 'Todo MVC Service'
    end

    server.mount '/todo/list', TodoListHandler
    server.mount '/todo', TodoHandler

    server
end